### PR TITLE
GEODE-7134: Replace bucket Set with bucket int[]

### PR DESCRIPTION
**Pull request by mivanac**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [*] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [*] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [*] Is your initial contribution a single, squashed commit?

- [*] Does `gradlew build` run cleanly?

- [*] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by pivotal-jbarrett** [comment]
Please update your PR commits with properly formatted and descriptive commit messages. See https://cwiki.apache.org/confluence/display/GEODE/Commit+Message+Format

**Comment by mivanac** [comment]
Bucket Set is replaced with bucket int[], where index 0 contains length of array.

**Comment by pivotal-jbarrett** [comment]
@mivanac I know this is draft but I think some input could guide. Perhaps so we can distinguish any arbitrary int[] from a bucket set maybe BucketSet should encapsulate this int[]? Is the rational for just int[] to save the extra allocation of the BucketSet object? If so maybe a set of static methods on a BucketSet helper class to encapsulate some of the logic, like length, iteration, set operations, such that no operations are done directly on the int[] but only through the helper. 
```
class BucketSet {
static int get(int[] bucketSet, int index) {
   return bucketSet[I+1];
}

static int length(int[] bucketSet) {
  return null == bucketSet || bucketSet.length < 2 ? 0 : bucketSet[0];
}

}

**Comment by mivanac** [comment]
Thanks for advice.

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+public class BucketSetHelper {
+  public static int get(int[] bucketSet, int index) {
+    return bucketSet[index + 1];
+  }
+
+  public static int length(int[] bucketSet) {
+    return null == bucketSet || bucketSet.length < 2 ? 0 : bucketSet[0];
+  }
+
+  public static void add(int[] bucketSet, int value) {
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java]: What if the bucketSet doesn't have capacity? Should it resize and return a new array?

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    int index = bucketSet[0] + 1;
+    bucketSet[index] = value;
+    bucketSet[0] = index;
+  }
+
+  public static Set<Integer> toSet(int[] bucketArray) {
+    Set<Integer> bucketSet;
+    if (BucketSetHelper.length(bucketArray) > 0) {
+      bucketSet =
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java]: How frequently is this being invoked? This is really expensive in time and allocations. I'd implement a more direct conversion, `new HashSet(length(bucketSet))`, then iterator over the `bucketSet` array and add each element to the `HashSet`.

What about just wrapping the current array in a Set implementation?

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+          new HashSet(
+              Arrays.asList(
+                  ArrayUtils.toObject(Arrays.copyOfRange(bucketArray, 1, bucketArray[0] + 1))));
+    } else {
+      bucketSet = new HashSet();
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java]: `Collections.emptySet()`?

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    }
+    return bucketSet;
+  }
+
+  public static int[] fromSet(Set<Integer> bucketSet) {
+    int[] bucketArray = new int[bucketSet.size() + 1];
+    bucketArray[0] = bucketSet.size();
+
+    if (bucketSet.size() > 0) {
+      System.arraycopy(ArrayUtils.toPrimitive(bucketSet.toArray(new Integer[bucketSet.size()])), 0,
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java]: Like above, this has extra allocations we don't need. I would iterate directly.

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java (hunk 1 above)
Comment (mivanac) [reply geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java]: Before add method is used, array is allocated, with enough capacity, so it should never happen that we don`t have capacity. If it happens, it is fault.

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java (hunk 4 above)
Comment (mivanac) [reply geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java]: Thanks for comments. Solution is updated.

**Comment by pivotal-jbarrett** [comment]
@mivanac It doesn't look like these test failures are your fault. Try merging the latest develop into this branch and pushing again.

**Comment by pivotal-jbarrett** [comment]
@mivanac It is your PR after all. Your local build should be failing too. The `Version` class and some of the internal serialization stuff have moved.

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  public static Set<Integer> toSet(int[] bucketSet) {
+    Set<Integer> resultSet;
+    int arrayLength = length(bucketSet);
+    if (arrayLength > 0) {
+      resultSet = new HashSet(arrayLength);
+      for (int i = 1; i <= arrayLength; i++) {
+        resultSet.add(bucketSet[i]);
+      }
+    } else {
+      resultSet = Collections.emptySet();
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java]: It looks like my recommendation to use `emptySet()` is causing some test failures. Looks like the consumer of this method expects to modify the set returned.
```
...
     final int arrayLength = length(bucketSet);
     final Set<Integer> resultSet = new HashSet(arrayLength);
     for (int i = 1; i <= arrayLength; i++) {
       resultSet.add(bucketSet[i]);
     }
     return resultSet;
...
```

File: geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+ * the License.
+ */
+package org.apache.geode.internal.cache;
+
+
+import java.util.HashSet;
+import java.util.Set;
+
+
+public class BucketSetHelper {
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/BucketSetHelper.java]: I think you should add a unit test for this new class. I know it is just a bunch of static methods, but you can have a unit test for them.

**Comment by pivotal-jbarrett** [comment]
@mivanac I don't think those test failures are yours but you should run them locally and if they pass write up some flaky test JIRAs for them if they don't already have them.

**Comment by mivanac** [comment]
Thanks for all the comments.

**Comment by mivanac** [comment]
Could you merge this PR, if there is nothing else to add.

**Pull request merged**